src/matrixrtc/NewMembershipManager.ts

```diff
@@ -0,0 +1,808 @@
+}
+
+interface Action {
+    /**
+     * When this action should be executed
+     */
+    ts: number;
+    /**
+     * The state of the different loops
+     * can also be thought of as the type of the action
+     */
+    type: DelayedLeaveActionType | MembershipActionType | DirectMembershipManagerActions;
+}
+
+/**
+ * @internal
+ */
+class ActionScheduler {
+    public state: ActionSchedulerState;
+
+    public constructor(
+        state: ActionSchedulerState,
+        private manager: Pick<MembershipManager, "delayedLeaveLoopHandler" | "membershipLoopHandler">,
+    ) {
+        this.state = state;
+    }
+    // state variables for a wakeup mechanism (in case we add some action externally and need to leave the current sleep)
+    private wakeupPromise?: Promise<void>;
+    private wakeup?: (value: void | PromiseLike<void>) => void;
+    private didWakeUp = false;
+
+    private actions: Action[] = [];
+    private insertions: Action[] = [];
+    private resetWith?: Action[];
+    /**
+     * This starts the main loop of the memberhsip manager that handles event sending, delayed event sending and delayed event restarting.
+     * @param initialActions The initial actions the manager will start with. It should be enough to pass: DelayedLeaveActionType.Initial
+     * @throws This throws an error only if the memberhsip cannot run anymore. For example it reached the maximum retires.
+     * In most other error cases the manager will try to handle any server errors by itself.
+     */
+    public async startWithActions(initialActions: Action[]): Promise<void> {
+        this.actions = initialActions;
+
+        while (this.actions.length > 0) {
+            this.actions.sort((a, b) => a.ts - b.ts);
+            logger.debug("Current MembershipManager action queue: ", this.actions, "\nDate.now: ", +Date.now());
+            const nextAction = this.actions[0];
+
+            this.wakeupPromise = new Promise((resolve) => {
+                this.wakeup = resolve;
+            });
+            if (nextAction.ts > Date.now()) await Promise.race([this.wakeupPromise, sleep(nextAction.ts - Date.now())]);
+            if (this.didWakeUp) {
+                // In case of a wakeup we do not want to run the next action because the next action now might be sth different.
+                // Instead we recompute the actions array and do another iteration.
+                this.didWakeUp = false;
+            } else {
+                try {
+                    if (isDelayedLeaveActionType(nextAction.type)) {
+                        await this.manager.delayedLeaveLoopHandler(this.state, nextAction.type);
+                    } else if (isMembershipActionType(nextAction.type)) {
+                        await this.manager.membershipLoopHandler(this.state, nextAction.type);
+                    }
+                } catch (e) {
+                    throw Error("The MemberhsipManager has to shut down because of the end condition: " + e);
```

Comment:
A few thoughts;

- I'm surprised we aren't doing any cleanup of the membership state
- If the intention is that this will only ever be reached when an explicit "end condition" throw was made, then should we be checking the type of error and only terminate if it was of that type
- In the current implementations of the loop handlers there aren't general purpose catch blocks, so I think any unanticipated error would hit this catch block

Reply:
There now is `scheduler.resetState()`

Reply:
```suggestion
                    throw Error("The MembershipManager has to shut down because of the end condition: " + e);
```